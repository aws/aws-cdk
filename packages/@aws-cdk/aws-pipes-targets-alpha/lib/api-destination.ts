import { IInputTransformation, IPipe, ITarget, TargetConfig } from '@aws-cdk/aws-pipes-alpha';
import { IApiDestination } from 'aws-cdk-lib/aws-events';
import { IRole, PolicyStatement } from 'aws-cdk-lib/aws-iam';

/**
 * EventBridge API destination target properties.
 */
export interface ApiDestinationTargetParameters {
  /**
   * The input transformation to apply to the message before sending it to the target.
   *
   * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-pipes-pipe-pipetargetparameters.html#cfn-pipes-pipe-pipetargetparameters-inputtemplate
   * @default - none
   */
  readonly inputTransformation?: IInputTransformation;

  /**
   * The headers to send as part of the request invoking the EventBridge API destination.
   *
   * The headers are merged with the headers from the API destination.
   * If there are conflicts, the headers from the API destination take precedence.
   *
   * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-pipes-pipe-pipetargethttpparameters.html#cfn-pipes-pipe-pipetargethttpparameters-headerparameters
   * @default - none
   */
  readonly headerParameters?: { [key: string]: string };

  /**
   * The path parameter values used to populate the EventBridge API destination path wildcards ("*").
   *
   * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-pipes-pipe-pipetargethttpparameters.html#cfn-pipes-pipe-pipetargethttpparameters-pathparametervalues
   * @default - none
   */
  readonly pathParameterValues?: string[];

  /**
   * The query string keys/values that need to be sent as part of request invoking the EventBridge API destination.
   *
   * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-pipes-pipe-pipetargethttpparameters.html#cfn-pipes-pipe-pipetargethttpparameters-querystringparameters
   * @default - none
   */
  readonly queryStringParameters?: Record<string, string>;
}

/**
 * A EventBridge Pipes target that sends messages to an EventBridge API destination.
 */
export class ApiDestinationTarget implements ITarget {
  private destination: IApiDestination;
  private apiParameters?: ApiDestinationTargetParameters;
  public readonly targetArn: string;

  constructor(destination: IApiDestination, parameters?: ApiDestinationTargetParameters) {
    this.destination = destination;
    this.apiParameters = parameters;
    this.targetArn = destination.apiDestinationArn;
  }

  grantPush(grantee: IRole): void {
    grantee.addToPrincipalPolicy(new PolicyStatement({
      resources: [this.destination.apiDestinationArn],
      actions: ['events:InvokeApiDestination'],
    }));
  }

  bind(pipe: IPipe): TargetConfig {
    if (!this.apiParameters) {
      return {
        targetParameters: {},
      };
    }

    return {
      targetParameters: {
        inputTemplate: this.apiParameters.inputTransformation?.bind(pipe).inputTemplate,
        httpParameters: this.apiParameters,
      },
    };
  }
}
