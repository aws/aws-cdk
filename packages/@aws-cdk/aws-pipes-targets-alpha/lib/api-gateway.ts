import { IInputTransformation, IPipe, ITarget, TargetConfig } from '@aws-cdk/aws-pipes-alpha';
import { IRestApi } from 'aws-cdk-lib/aws-apigateway';
import { IRole, PolicyStatement } from 'aws-cdk-lib/aws-iam';

/**
 * API Gateway REST API target properties.
 */
export interface ApiGatewayTargetParameters {
  /**
   * The input transformation to apply to the message before sending it to the target.
   *
   * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-pipes-pipe-pipetargetparameters.html#cfn-pipes-pipe-pipetargetparameters-inputtemplate
   * @default - none
   */
  readonly inputTransformation?: IInputTransformation;

  /**
   * The method for API Gateway resource.
   *
   * @default '*' - ANY
   */
  readonly method?: string;

  /**
   * The path for the API Gateway resource.
   *
   * @default '/'
   */
  readonly path?: string;

  /**
   * The deployment stage for the API Gateway resource.
   *
   * @default - the value of `deploymentStage.stageName` of target API Gateway resource.
   */
  readonly stage?: string;

  /**
   * The headers to send as part of the request invoking the API Gateway REST API.
   *
   * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-pipes-pipe-pipetargethttpparameters.html#cfn-pipes-pipe-pipetargethttpparameters-headerparameters
   * @default - none
   */
  readonly headerParameters?: Record<string, string>;

  /**
   * The path parameter values used to populate the API Gateway REST API path wildcards ("*").
   *
   * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-pipes-pipe-pipetargethttpparameters.html#cfn-pipes-pipe-pipetargethttpparameters-pathparametervalues
   * @default - none
   */
  readonly pathParameterValues?: string[];

  /**
   * The query string keys/values that need to be sent as part of request invoking the API Gateway REST API.
   *
   * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-pipes-pipe-pipetargethttpparameters.html#cfn-pipes-pipe-pipetargethttpparameters-querystringparameters
   * @default - none
   */
  readonly queryStringParameters?: Record<string, string>;
}

/**
 * An EventBridge Pipes target that sends messages to an EventBridge API destination.
 */
export class ApiGatewayTarget implements ITarget {
  private restApi: IRestApi;
  private restApiParameters?: ApiGatewayTargetParameters;
  private restApiArn: string;
  public readonly targetArn;

  constructor(restApi: IRestApi, parameters?: ApiGatewayTargetParameters) {
    this.restApi = restApi;
    this.restApiParameters = parameters;

    if (this.restApiParameters?.stage === undefined && this.restApi.deploymentStage === undefined) {
      throw Error('The REST API must have a deployed stage.');
    }

    this.restApiArn = this.restApi.arnForExecuteApi(
      this.restApiParameters?.method,
      this.restApiParameters?.path || '/',
      this.restApiParameters?.stage || this.restApi.deploymentStage.stageName,
    );

    this.targetArn = this.restApiArn;
  }

  grantPush(grantee: IRole): void {
    grantee.addToPrincipalPolicy(new PolicyStatement({
      resources: [this.restApiArn],
      actions: ['execute-api:Invoke'],
    }));
  }

  bind(pipe: IPipe): TargetConfig {
    if (!this.restApiParameters) {
      return {
        targetParameters: {},
      };
    }

    return {
      targetParameters: {
        inputTemplate: this.restApiParameters.inputTransformation?.bind(pipe).inputTemplate,
        httpParameters: this.restApiParameters,
      },
    };
  }
}
