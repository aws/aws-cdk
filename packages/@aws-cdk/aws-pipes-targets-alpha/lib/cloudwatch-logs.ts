import { IInputTransformation, IPipe, ITarget, TargetConfig } from '@aws-cdk/aws-pipes-alpha';
import { Token } from 'aws-cdk-lib';
import { IRole } from 'aws-cdk-lib/aws-iam';
import { ILogGroup } from 'aws-cdk-lib/aws-logs';

/**
 * CloudWatch Logs target properties.
 */
export interface CloudWatchLogsTargetParameters {
  /**
   * The input transformation to apply to the message before sending it to the target.
   *
   * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-pipes-pipe-pipetargetparameters.html#cfn-pipes-pipe-pipetargetparameters-inputtemplate
   * @default - none
   */
  readonly inputTransformation?: IInputTransformation;

  /**
   * The name of the log stream.
   *
   * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-pipes-pipe-pipetargetcloudwatchlogsparameters.html#cfn-pipes-pipe-pipetargetcloudwatchlogsparameters-logstreamname
   * @default - none
   */
  readonly logStreamName?: string;

  /**
   * The JSON path expression that references the timestamp in the payload.
   * This is the time that the event occurred, as a JSON path expression in the payload.
   *
   * @example '$.data.timestamp'
   *
   * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-pipes-pipe-pipetargetcloudwatchlogsparameters.html#cfn-pipes-pipe-pipetargetcloudwatchlogsparameters-timestamp
   * @default - current time
   */
  readonly timestamp?: string;
}

/**
 * An EventBridge Pipes target that sends messages to a CloudWatch Logs log group.
 */
export class CloudWatchLogsTarget implements ITarget {
  private logGroup: ILogGroup;
  private cloudWatchLogsParameters?: CloudWatchLogsTargetParameters;
  public readonly targetArn: string;

  constructor(logGroup: ILogGroup, parameters?: CloudWatchLogsTargetParameters) {
    this.logGroup = logGroup;
    this.targetArn = logGroup.logGroupArn;

    if (parameters) {
      this.cloudWatchLogsParameters = parameters;
      for (const validate of [validateLogStreamName, validateTimestamp]) {
        validate(parameters);
      }
    }
  }

  grantPush(grantee: IRole): void {
    this.logGroup.grantWrite(grantee);
  }

  bind(pipe: IPipe): TargetConfig {
    if (!this.cloudWatchLogsParameters) {
      return {
        targetParameters: {},
      };
    }

    return {
      targetParameters: {
        inputTemplate: this.cloudWatchLogsParameters.inputTransformation?.bind(pipe).inputTemplate,
        cloudWatchLogsParameters: this.cloudWatchLogsParameters,
      },
    };
  }
}

function validateLogStreamName({ logStreamName }: CloudWatchLogsTargetParameters) {
  if (logStreamName !== undefined && !Token.isUnresolved(logStreamName)) {
    if (logStreamName.length < 1 || logStreamName.length > 256) {
      throw new Error(`Log stream name must be between 1 and 256 characters, received ${logStreamName.length}`);
    }
  }
}

function validateTimestamp({ timestamp }: CloudWatchLogsTargetParameters) {
  if (timestamp !== undefined && !Token.isUnresolved(timestamp)) {
    if (timestamp.length < 1 || timestamp.length > 256) {
      throw new Error(`Timestamp must be between 1 and 256 characters, received ${timestamp.length}`);
    }
  }
}
