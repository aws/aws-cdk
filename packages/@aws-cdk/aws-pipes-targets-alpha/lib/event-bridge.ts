import { IInputTransformation, IPipe, ITarget, TargetConfig } from '@aws-cdk/aws-pipes-alpha';
import { Token } from 'aws-cdk-lib';
import { IEventBus } from 'aws-cdk-lib/aws-events';
import { IRole } from 'aws-cdk-lib/aws-iam';

/**
 * EventBridge target properties.
 */
export interface EventBridgeTargetParameters {
  /**
   * The input transformation to apply to the message before sending it to the target.
   *
   * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-pipes-pipe-pipetargetparameters.html#cfn-pipes-pipe-pipetargetparameters-inputtemplate
   * @default - none
   */
  readonly inputTransformation?: IInputTransformation;

  /**
   * A free-form string used to decide what fields to expect in the event detail.
   *
   * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-pipes-pipe-pipetargeteventbridgeeventbusparameters.html#cfn-pipes-pipe-pipetargeteventbridgeeventbusparameters-detailtype
   * @default - none
   */
  readonly detailType?: string;

  /**
   * The URL subdomain of the endpoint.
   *
   * @example
   * // if the URL for the endpoint is https://abcde.veo.endpoints.event.amazonaws.com
   * 'abcde.veo'
   *
   * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-pipes-pipe-pipetargeteventbridgeeventbusparameters.html#cfn-pipes-pipe-pipetargeteventbridgeeventbusparameters-endpointid
   * @default - none
   */
  readonly endpointId?: string;

  /**
   * AWS resources, identified by Amazon Resource Name (ARN), which the event primarily concerns.
   *
   * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-pipes-pipe-pipetargeteventbridgeeventbusparameters.html#cfn-pipes-pipe-pipetargeteventbridgeeventbusparameters-resources
   * @default - none
   */
  readonly resources?: string[];

  /**
   * The source of the event.
   *
   * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-pipes-pipe-pipetargeteventbridgeeventbusparameters.html#cfn-pipes-pipe-pipetargeteventbridgeeventbusparameters-source
   * @default - none
   */
  readonly source?: string;

  /**
   * The time stamp of the event, per RFC3339.
   *
   * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-pipes-pipe-pipetargeteventbridgeeventbusparameters.html#cfn-pipes-pipe-pipetargeteventbridgeeventbusparameters-time
   * @default - the time stamp of the PutEvents call
   */
  readonly time?: string;
}

/**
 * An EventBridge Pipes target that sends messages to an EventBridge event bus.
 */
export class EventBridgeTarget implements ITarget {
  private eventBus: IEventBus;
  private eventBridgeParameters?: EventBridgeTargetParameters;
  public readonly targetArn: string;

  constructor(eventBus: IEventBus, parameters?: EventBridgeTargetParameters) {
    this.eventBus = eventBus;
    this.targetArn = eventBus.eventBusArn;
    if (parameters) {
      this.eventBridgeParameters = parameters;
      for (const validate of [validateDetailType, validateEndpointId, validateSource, validateTime]) {
        validate(parameters);
      }
    }
  }

  grantPush(grantee: IRole): void {
    this.eventBus.grantPutEventsTo(grantee);
  }

  bind(pipe: IPipe): TargetConfig {
    if (!this.eventBridgeParameters) {
      return {
        targetParameters: {},
      };
    }

    return {
      targetParameters: {
        inputTemplate: this.eventBridgeParameters.inputTransformation?.bind(pipe).inputTemplate,
        eventBridgeEventBusParameters: this.eventBridgeParameters,
      },
    };
  }
}

function validateDetailType({ detailType }: EventBridgeTargetParameters) {
  if (detailType !== undefined && !Token.isUnresolved(detailType)) {
    if (detailType.length < 1 || detailType.length > 128) {
      throw new Error(`Detail type must be between 1 and 128 characters, received ${detailType.length}`);
    }
  }
}

function validateEndpointId({ endpointId }: EventBridgeTargetParameters) {
  if (endpointId !== undefined && !Token.isUnresolved(endpointId)) {
    if (endpointId.length < 1 || endpointId.length > 50) {
      throw new Error(`Endpoint id must be between 1 and 50 characters, received ${endpointId.length}`);
    }
  }
}

function validateSource({ source }: EventBridgeTargetParameters) {
  if (source !== undefined && !Token.isUnresolved(source)) {
    if (source.length < 1 || source.length > 256) {
      throw new Error(`Source must be between 1 and 256 characters, received ${source.length}`);
    }
  }
}

function validateTime({ time }: EventBridgeTargetParameters) {
  if (time !== undefined && !Token.isUnresolved(time)) {
    if (time.length < 1 || time.length > 256) {
      throw new Error(`Time must be between 1 and 256 characters, received ${time.length}`);
    }
  }
}
