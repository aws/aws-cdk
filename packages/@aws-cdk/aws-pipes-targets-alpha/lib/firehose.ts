import { IInputTransformation, IPipe, ITarget, TargetConfig } from '@aws-cdk/aws-pipes-alpha';
import { IRole } from 'aws-cdk-lib/aws-iam';
import { IDeliveryStream } from 'aws-cdk-lib/aws-kinesisfirehose';

/**
 * Amazon Data Firehose target properties.
 */
export interface FirehoseTargetParameters {
  /**
   * The input transformation to apply to the message before sending it to the target.
   *
   * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-pipes-pipe-pipetargetparameters.html#cfn-pipes-pipe-pipetargetparameters-inputtemplate
   * @default - none
   */
  readonly inputTransformation?: IInputTransformation;
}

/**
 * An EventBridge Pipes target that sends messages to an Amazon Data Firehose delivery stream.
 */
export class FirehoseTarget implements ITarget {
  private deliveryStream: IDeliveryStream;
  private deliveryStreamParameters: FirehoseTargetParameters;
  public readonly targetArn: string;

  constructor(deliveryStream: IDeliveryStream, parameters: FirehoseTargetParameters = {}) {
    this.deliveryStream = deliveryStream;
    this.targetArn = deliveryStream.deliveryStreamArn;
    this.deliveryStreamParameters = parameters;
  }

  grantPush(grantee: IRole): void {
    this.deliveryStream.grantPutRecords(grantee);
  }

  bind(pipe: IPipe): TargetConfig {
    return {
      targetParameters: {
        inputTemplate: this.deliveryStreamParameters.inputTransformation?.bind(pipe).inputTemplate,
      },
    };
  }
}
