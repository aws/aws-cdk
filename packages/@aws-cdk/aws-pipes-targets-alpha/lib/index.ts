export * from './sqs';
export * from './stepfunctions';
