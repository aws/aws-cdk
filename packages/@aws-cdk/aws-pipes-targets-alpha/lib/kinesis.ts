import { IInputTransformation, IPipe, ITarget, TargetConfig } from '@aws-cdk/aws-pipes-alpha';
import { Token } from 'aws-cdk-lib';
import { IRole } from 'aws-cdk-lib/aws-iam';
import { IStream } from 'aws-cdk-lib/aws-kinesis';

/**
 * Kinesis target properties.
 */
export interface KinesisTargetParameters {
  /**
   * The input transformation to apply to the message before sending it to the target.
   *
   * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-pipes-pipe-pipetargetparameters.html#cfn-pipes-pipe-pipetargetparameters-inputtemplate
   * @default - none
   */
  readonly inputTransformation?: IInputTransformation;

  /**
   * Determines which shard in the stream the data record is assigned to.
   *
   * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-pipes-pipe-pipetargetkinesisstreamparameters.html#cfn-pipes-pipe-pipetargetkinesisstreamparameters-partitionkey
   */
  readonly partitionKey: string;
}

/**
 * An EventBridge Pipes target that sends messages to a Kinesis stream.
 */
export class KinesisTarget implements ITarget {
  private stream: IStream;
  private streamParameters: KinesisTargetParameters;
  public readonly targetArn: string;

  constructor(stream: IStream, parameters: KinesisTargetParameters) {
    this.stream = stream;
    this.targetArn = stream.streamArn;
    this.streamParameters = parameters;

    validatePartitionKey(parameters.partitionKey);
  }

  grantPush(grantee: IRole): void {
    this.stream.grantWrite(grantee);
  }

  bind(pipe: IPipe): TargetConfig {
    return {
      targetParameters: {
        inputTemplate: this.streamParameters.inputTransformation?.bind(pipe).inputTemplate,
        kinesisStreamParameters: this.streamParameters,
      },
    };
  }
}

function validatePartitionKey(pk: string) {
  if (!Token.isUnresolved(pk) && pk.length > 256) {
    throw new Error(`Partition key must be less than or equal to 256 characters, received ${pk.length}`);
  }
}
