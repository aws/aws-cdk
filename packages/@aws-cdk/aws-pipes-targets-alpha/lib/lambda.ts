import { IInputTransformation, IPipe, ITarget, TargetConfig } from '@aws-cdk/aws-pipes-alpha';
import { IRole } from 'aws-cdk-lib/aws-iam';
import * as lambda from 'aws-cdk-lib/aws-lambda';

/**
 * Parameters for the LambdaFunction target
 */
export interface LambdaFunctionParameters {

  /**
   * The input transformation to apply to the message before sending it to the target.
   *
   * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-pipes-pipe-pipetargetparameters.html#cfn-pipes-pipe-pipetargetparameters-inputtemplate
   * @default - none
   */
  readonly inputTransformation?: IInputTransformation;

  /**
   * Specify whether to invoke the Lambda Function synchronously (`REQUEST_RESPONSE`) or asynchronously (`FIRE_AND_FORGET`).
   *
   * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-pipes-pipe-pipetargetlambdafunctionparameters.html
   * @default LambdaFunctionInvocationType.REQUEST_RESPONSE
   */
  readonly invocationType?: LambdaFunctionInvocationType;
}

/**
 * InvocationType for invoking the Lambda Function.
 * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-pipes-pipe-pipetargetlambdafunctionparameters.html
 */
export enum LambdaFunctionInvocationType {
  /**
   * Invoke Lambda Function asynchronously (`Invoke`). `InvocationType` is set to `Event` on `Invoke`, see https://docs.aws.amazon.com/lambda/latest/api/API_Invoke.html for more details.
   */
  FIRE_AND_FORGET = 'FIRE_AND_FORGET',

  /**
   * Invoke Lambda Function synchronously (`Invoke`) and wait for the response. `InvocationType` is set to `RequestResponse` on `Invoke`, see https://docs.aws.amazon.com/lambda/latest/api/API_Invoke.html for more details.
   */
  REQUEST_RESPONSE = 'REQUEST_RESPONSE',
}

/**
 * An EventBridge Pipes target that sends messages to an AWS Lambda Function.
 */
export class LambdaFunction implements ITarget {
  public readonly targetArn: string;

  private readonly lambdaFunction: lambda.IFunction;
  private readonly invocationType: LambdaFunctionInvocationType;
  private readonly inputTemplate?: IInputTransformation;

  constructor(
    lambdaFunction: lambda.IFunction,
    parameters: LambdaFunctionParameters,
  ) {
    this.lambdaFunction = lambdaFunction;
    this.targetArn = lambdaFunction.functionArn;
    this.invocationType =
      parameters.invocationType ??
      LambdaFunctionInvocationType.REQUEST_RESPONSE;
    this.inputTemplate = parameters.inputTransformation;
  }

  grantPush(grantee: IRole): void {
    this.lambdaFunction.grantInvoke(grantee);
  }

  bind(pipe: IPipe): TargetConfig {
    return {
      targetParameters: {
        inputTemplate: this.inputTemplate?.bind(pipe).inputTemplate,
        lambdaFunctionParameters: {
          invocationType: this.invocationType,
        },
      },
    };
  }
}
