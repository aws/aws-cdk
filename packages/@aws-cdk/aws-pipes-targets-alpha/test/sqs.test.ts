import { InputTransformation, Pipe } from '@aws-cdk/aws-pipes-alpha';
import { App, Stack } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { Queue } from 'aws-cdk-lib/aws-sqs';
import { TestSource } from './test-classes';
import { SqsTarget } from '../lib';

describe('sqs', () => {
  it('should have only target arn', () => {
    // ARRANGE
    const app = new App();
    const stack = new Stack(app, 'TestStack');

    const queue = new Queue(stack, 'MySqs', {});

    const target = new SqsTarget(queue);

    new Pipe(stack, 'MyPipe', {
      source: new TestSource(),
      target,
    });

    // ACT
    const template = Template.fromStack(stack);

    // ASSERT
    template.hasResourceProperties('AWS::Pipes::Pipe', {
      Target: {
        'Fn::GetAtt': [
          'MySqs4F2D580E',
          'Arn',
        ],
      },
      TargetParameters: {},
    });
  });

  it('should have target parameters', () => {
    // ARRANGE
    const app = new App();
    const stack = new Stack(app, 'TestStack');

    const queue = new Queue(stack, 'MySqs', {});

    const target = new SqsTarget(queue, {
      messageDeduplicationId: 'deduplication-id',
      messageGroupId: 'group-id',
    });

    new Pipe(stack, 'MyPipe', {
      source: new TestSource(),
      target,
    });

    // ACT
    const template = Template.fromStack(stack);

    // ASSERT
    template.hasResourceProperties('AWS::Pipes::Pipe', {
      TargetParameters: {
        SqsQueueParameters: {
          MessageDeduplicationId: 'deduplication-id',
          MessageGroupId: 'group-id',
        },
      },
    });
  });

  it('should have input transformation', () => {
    // ARRANGE
    const app = new App();
    const stack = new Stack(app, 'TestStack');

    const queue = new Queue(stack, 'MySqs', {});

    const inputTransformation = InputTransformation.fromObject({
      key: 'value',
    });
    const target = new SqsTarget(queue, {
      inputTransformation,
    });

    new Pipe(stack, 'MyPipe', {
      source: new TestSource(),
      target,
    });

    // ACT
    const template = Template.fromStack(stack);

    // ASSERT
    template.hasResourceProperties('AWS::Pipes::Pipe', {
      TargetParameters: {
        InputTemplate: '{"key":"value"}',
      },
    });
  });

  it('should grant pipe role push access', () => {
    // ARRANGE
    const app = new App();
    const stack = new Stack(app, 'TestStack');

    const queue = new Queue(stack, 'MySqs', {});

    const target = new SqsTarget(queue);

    new Pipe(stack, 'MyPipe', {
      source: new TestSource(),
      target,
    });

    // ACT
    const template = Template.fromStack(stack);

    // ASSERT
    expect(template.findResources('AWS::IAM::Role')).toMatchSnapshot();
    expect(template.findResources('AWS::IAM::Policy')).toMatchSnapshot();
  });
});
