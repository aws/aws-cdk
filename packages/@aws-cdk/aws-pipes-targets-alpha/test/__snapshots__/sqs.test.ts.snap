// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sqs should grant pipe role push access 1`] = `
{
  "MyPipeRoleCBC8E9AB": {
    "Properties": {
      "AssumeRolePolicyDocument": {
        "Statement": [
          {
            "Action": "sts:AssumeRole",
            "Effect": "Allow",
            "Principal": {
              "Service": "pipes.amazonaws.com",
            },
          },
        ],
        "Version": "2012-10-17",
      },
    },
    "Type": "AWS::IAM::Role",
  },
}
`;

exports[`sqs should grant pipe role push access 2`] = `
{
  "MyPipeRoleDefaultPolicy31387C20": {
    "Properties": {
      "PolicyDocument": {
        "Statement": [
          {
            "Action": [
              "sqs:SendMessage",
              "sqs:GetQueueAttributes",
              "sqs:GetQueueUrl",
            ],
            "Effect": "Allow",
            "Resource": {
              "Fn::GetAtt": [
                "MySqs4F2D580E",
                "Arn",
              ],
            },
          },
        ],
        "Version": "2012-10-17",
      },
      "PolicyName": "MyPipeRoleDefaultPolicy31387C20",
      "Roles": [
        {
          "Ref": "MyPipeRoleCBC8E9AB",
        },
      ],
    },
    "Type": "AWS::IAM::Policy",
  },
}
`;
